[ERROR] panic error in klc-controller-manager pod

Hello keptn community hope everything is going great,

Recently I installed keptn lifecycle toolkit and everything going great, I was testing some pre deployments evaluations but suddenly today I have an issue with the klc controller manager pod it’s a panic error of what I can see in my app keptnworkloadinstance

`1.6715559396021056e+09  INFO    KeptnTaskDefinition Controller  updated configmap status reference for: precheck
1.6715559396021402e+09  INFO    KeptnTaskDefinition Controller  Finished Reconciling KeptnTaskDefinition
1.671555939647164e+09   INFO    Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference        {"controller": "keptnworkloadinstance", "controllerGroup": "lifecycle.keptn.sh", "controllerKind": "KeptnWorkloadInstance", "KeptnWorkloadInstance": {"name":"keptn-grpc-svc-grpc-service-dev-1.0","namespace":"keptn-dev"}, "namespace": "keptn-dev", "name": "keptn-grpc-svc-grpc-service-dev-1.0", "reconcileID": "66209f97-4279-47b1-8c01-b2e382699b7c"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x145e0ab]

goroutine 421 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x16bb040, 0x26e9390})
        /usr/local/go/src/runtime/panic.go:838 +0x207
github.com/keptn/lifecycle-toolkit/operator/controllers/keptnworkloadinstance.(*KeptnWorkloadInstanceReconciler).getDesiredReplicas(0xc0002c0a80, {0x1b439e8, 0xc0005ee6f0}, {{0xc0013786c0, 0x14}, {0xc001368e60, 0x7}, {0xc00135d260, 0x1b}, {0xc001376720, ...}, ...}, ...)
        /workspace/controllers/keptnworkloadinstance/reconcile_deploymentstate.go:101 +0x1eb
github.com/keptn/lifecycle-toolkit/operator/controllers/keptnworkloadinstance.(*KeptnWorkloadInstanceReconciler).isReplicaSetRunning(0xc0002c0a80, {0x1b439e8, 0xc0005ee6f0}, {{0xc000758690?, 0x7f21186297f8?}, {0xc0002afe00?, 0x18ecfa7?}}, {0xc0002afdb6, 0x9})
        /workspace/controllers/keptnworkloadinstance/reconcile_deploymentstate.go:52 +0x285
github.com/keptn/lifecycle-toolkit/operator/controllers/keptnworkloadinstance.(*KeptnWorkloadInstanceReconciler).reconcileDeployment(0xc0002c0a80, {0x1b439e8, 0xc0005ee6f0}, 0xc00086c000)
        /workspace/controllers/keptnworkloadinstance/reconcile_deploymentstate.go:27 +0x157
github.com/keptn/lifecycle-toolkit/operator/controllers/keptnworkloadinstance.(*KeptnWorkloadInstanceReconciler).Reconcile.func4()
        /workspace/controllers/keptnworkloadinstance/controller.go:199 +0x29
github.com/keptn/lifecycle-toolkit/operator/controllers/common.PhaseHandler.HandlePhase({{0x1b490d0, 0xc00012a870}, {0x1b427a0, 0xc000780840}, {{0x1b45818, 0xc0005ef3e0}, 0x0}, 0xc00052b080}, {0x1b439e8, 0xc0005ee6f0}, ...)
        /workspace/controllers/common/phasehandler.go:49 +0x2c8
github.com/keptn/lifecycle-toolkit/operator/controllers/keptnworkloadinstance.(*KeptnWorkloadInstanceReconciler).Reconcile(0xc0002c0a80, {0x1b439e8, 0xc0005ee570}, {{{0xc0002afdb6?, 0x9?}, {0xc000758600?, 0x23?}}})
        /workspace/controllers/keptnworkloadinstance/controller.go:201 +0x1288
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1b43940?, {0x1b439e8?, 0xc0005ee570?}, {{{0xc0002afdb6?, 0x17f7f00?}, {0xc000758600?, 0x4041f4?}}})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0004c7a40, {0x1b43940, 0xc0004f9c00}, {0x171d260?, 0xc0004a8040?})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0004c7a40, {0x1b43940, 0xc0004f9c00})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:230 +0x325`

If I check the workloadinstance it seems ok 

` Normal  AppPreDeployEvaluationsFinishedSuccess  10m    keptnworkloadinstance-controller  App Pre-Deployment Evaluations Pre evaluations tasks for app have finished successfully / Namespace: keptn-dev, Name: keptn-grpc-svc-grpc-service-dev-1.0, Version: 1.0
  Normal  AppPreDeployEvaluationsFinishedSuccess  7m41s  keptnworkloadinstance-controller  App Pre-Deployment Evaluations Pre evaluations tasks for app have finished successfully / Namespace: keptn-dev, Name: keptn-grpc-svc-grpc-service-dev-1.0, Version: 1.0
  Normal  AppPreDeployEvaluationsFinishedSuccess  2m17s  keptnworkloadinstance-controller  App Pre-Deployment Evaluations Pre evaluations tasks for app have finished successfully / Namespace: keptn-dev, Name: keptn-grpc-svc-grpc-service-dev-1.0, Version: 1.0`

Any help will be appreciated thank you very much

--- a/operator/controllers/keptnworkloadinstance/controller_test.go
+++ b/operator/controllers/keptnworkloadinstance/controller_test.go
@@ -3,6 +3,7 @@ package keptnworkloadinstance
 import (
 	"context"
 	"fmt"
+	controllererrors "github.com/keptn/lifecycle-toolkit/operator/controllers/errors"
 	"strings"
 	"testing"
 
@@ -53,23 +54,29 @@ func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_FailedReplicaSet(t 
 	testrequire.Equal(t, apicommon.StateProgressing, keptnState)
 }
 
-func makeWorkloadInstanceWithRef(objectMeta metav1.ObjectMeta, refKind string) *klcv1alpha2.KeptnWorkloadInstance {
-	workloadInstance := &klcv1alpha2.KeptnWorkloadInstance{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "my-wli",
-			Namespace: "default",
-		},
-		Spec: klcv1alpha2.KeptnWorkloadInstanceSpec{
-			KeptnWorkloadSpec: klcv1alpha2.KeptnWorkloadSpec{
-				ResourceReference: klcv1alpha2.ResourceReference{
-					UID:  objectMeta.UID,
-					Name: objectMeta.Name,
-					Kind: refKind,
-				},
-			},
-		},
+func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_UnavailableReplicaSet(t *testing.T) {
+
+	rep := int32(1)
+	replicasetFail := makeReplicaSet("myrep", "default", &rep, 0)
+
+	// do not put the ReplicaSet into the cluster
+	fakeClient := k8sfake.NewClientBuilder().WithObjects().Build()
+
+	err := klcv1alpha2.AddToScheme(fakeClient.Scheme())
+	testrequire.Nil(t, err)
+
+	workloadInstance := makeWorkloadInstanceWithRef(replicasetFail.ObjectMeta, "ReplicaSet")
+
+	err = fakeClient.Create(context.TODO(), workloadInstance)
+	require.Nil(t, err)
+
+	r := &KeptnWorkloadInstanceReconciler{
+		Client: fakeClient,
 	}
-	return workloadInstance
+
+	keptnState, err := r.reconcileDeployment(context.TODO(), workloadInstance)
+	testrequire.NotNil(t, err)
+	testrequire.Equal(t, apicommon.StateUnknown, keptnState)
 }
 
 func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_FailedStatefulSet(t *testing.T) {
@@ -96,6 +103,31 @@ func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_FailedStatefulSet(t
 	testrequire.Equal(t, apicommon.StateProgressing, keptnState)
 }
 
+func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_UnavailableStatefulSet(t *testing.T) {
+
+	rep := int32(1)
+	statefulSetFail := makeStatefulSet("mystat", "default", &rep, 0)
+
+	// do not put the StatefulSet into the cluster
+	fakeClient := k8sfake.NewClientBuilder().WithObjects().Build()
+
+	err := klcv1alpha2.AddToScheme(fakeClient.Scheme())
+	testrequire.Nil(t, err)
+
+	workloadInstance := makeWorkloadInstanceWithRef(statefulSetFail.ObjectMeta, "StatefulSet")
+
+	err = fakeClient.Create(context.TODO(), workloadInstance)
+	require.Nil(t, err)
+
+	r := &KeptnWorkloadInstanceReconciler{
+		Client: fakeClient,
+	}
+
+	keptnState, err := r.reconcileDeployment(context.TODO(), workloadInstance)
+	testrequire.NotNil(t, err)
+	testrequire.Equal(t, apicommon.StateUnknown, keptnState)
+}
+
 func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_FailedDaemonSet(t *testing.T) {
 
 	daemonSetFail := makeDaemonSet("mystat", "default", 1, 0)
@@ -117,6 +149,29 @@ func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_FailedDaemonSet(t *
 	keptnState, err := r.reconcileDeployment(context.TODO(), workloadInstance)
 	testrequire.Nil(t, err)
 	testrequire.Equal(t, apicommon.StateProgressing, keptnState)
+}
+
+func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_UnavailableDaemonSet(t *testing.T) {
+	daemonSetFail := makeDaemonSet("mystat", "default", 1, 0)
+
+	// do not put the DaemonSet into the cluster
+	fakeClient := k8sfake.NewClientBuilder().WithObjects().Build()
+
+	err := klcv1alpha2.AddToScheme(fakeClient.Scheme())
+	testrequire.Nil(t, err)
+
+	workloadInstance := makeWorkloadInstanceWithRef(daemonSetFail.ObjectMeta, "DaemonSet")
+
+	err = fakeClient.Create(context.TODO(), workloadInstance)
+	require.Nil(t, err)
+
+	r := &KeptnWorkloadInstanceReconciler{
+		Client: fakeClient,
+	}
+
+	keptnState, err := r.reconcileDeployment(context.TODO(), workloadInstance)
+	testrequire.NotNil(t, err)
+	testrequire.Equal(t, apicommon.StateUnknown, keptnState)
 }
 
 func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_ReadyReplicaSet(t *testing.T) {
@@ -188,6 +243,27 @@ func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_ReadyDaemonSet(t *t
 	keptnState, err := r.reconcileDeployment(context.TODO(), workloadInstance)
 	testrequire.Nil(t, err)
 	testrequire.Equal(t, apicommon.StateSucceeded, keptnState)
+}
+
+func TestKeptnWorkloadInstanceReconciler_reconcileDeployment_UnsupportedReferenceKind(t *testing.T) {
+
+	fakeClient := k8sfake.NewClientBuilder().WithObjects().Build()
+
+	err := klcv1alpha2.AddToScheme(fakeClient.Scheme())
+	testrequire.Nil(t, err)
+
+	workloadInstance := makeWorkloadInstanceWithRef(metav1.ObjectMeta{}, "Unknown")
+
+	err = fakeClient.Create(context.TODO(), workloadInstance)
+	require.Nil(t, err)
+
+	r := &KeptnWorkloadInstanceReconciler{
+		Client: fakeClient,
+	}
+
+	keptnState, err := r.reconcileDeployment(context.TODO(), workloadInstance)
+	testrequire.ErrorIs(t, err, controllererrors.ErrUnsupportedWorkloadInstanceResourceReference)
+	testrequire.Equal(t, apicommon.StateUnknown, keptnState)
 }
 
 func TestKeptnWorkloadInstanceReconciler_IsPodRunning(t *testing.T) {
@@ -879,4 +955,23 @@ func setupReconciler(t *testing.T) (*KeptnWorkloadInstanceReconciler, chan strin
 		SpanHandler: &controllercommon.SpanHandler{},
 	}
 	return r, recorder.Events, tr
+}
+
+func makeWorkloadInstanceWithRef(objectMeta metav1.ObjectMeta, refKind string) *klcv1alpha2.KeptnWorkloadInstance {
+	workloadInstance := &klcv1alpha2.KeptnWorkloadInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-wli",
+			Namespace: "default",
+		},
+		Spec: klcv1alpha2.KeptnWorkloadInstanceSpec{
+			KeptnWorkloadSpec: klcv1alpha2.KeptnWorkloadSpec{
+				ResourceReference: klcv1alpha2.ResourceReference{
+					UID:  objectMeta.UID,
+					Name: objectMeta.Name,
+					Kind: refKind,
+				},
+			},
+		},
+	}
+	return workloadInstance
 }

